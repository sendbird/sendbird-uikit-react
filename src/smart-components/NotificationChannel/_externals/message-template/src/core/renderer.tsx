import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FC<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface RendererOptions<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FC<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FC<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  opts?: RendererOptions<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: opts?.views?.box || FRAGMENT,
    text: opts?.views?.text || FRAGMENT,
    image: opts?.views?.image || FRAGMENT,
    imageButton: opts?.views?.imageButton || FRAGMENT,
    textButton: opts?.views?.textButton || FRAGMENT,
  };
}
