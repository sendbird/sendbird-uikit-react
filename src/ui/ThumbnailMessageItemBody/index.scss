@import '../../styles/variables';

.sendbird-thumbnail-message-item-body {
  position: relative;
  display: block;
  box-sizing: border-box;
  overflow: hidden;
  width: 100%;
  min-width: 360px;
  max-width: 400px;
  height: 270px;
  border-radius: 16px;
  &.reactions {
    border-radius: 16px 16px 0px 0px;
  }
  @include themed() { background-color: t(bg-1) }
  &.mouse-hover, &:hover {
    @include themed() { background-color: t(bg-2) }
  }

  .sendbird-thumbnail-message-item-body__thumbnail {
    position: absolute;
    border-radius: 16px;
  }

  .sendbird-thumbnail-message-item-body__placeholder {
    position: absolute;
    width: 100%;
    height: 270px;
    display: flex;
    justify-content: center;
    align-items: center;

    .sendbird-thumbnail-message-item-body__placeholder__icon {
      display: flex;
      align-items: center;
      justify-content: center;
      width: 56px;
      height: 56px;
      border-radius: 50%;
      background-color: var(--sendbird-light-background-50);
    }
  }

  .sendbird-thumbnail-message-item-body__image-cover {
    position: absolute;
    top: 0px;
    display: none;
    width: 100%;
    height: 270px;
    border-radius: 16px;;
    background-color: var(--sendbird-light-overlay-01);
  }

  .sendbird-thumbnail-message-item-body__video {
    position: absolute;
    width: 100%;
    height: 270px;
    border-radius: 16px;
  }

  .sendbird-thumbnail-message-item-body__icon-wrapper {
    position: absolute;
    width: 100%;
    height: 270px;
    display: flex;
    align-items: center;
    justify-content: center;

    .sendbird-thumbnail-message-item-body__icon-wrapper__icon {
      display: flex;
      align-items: center;
      justify-content: center;
      width: 56px;
      height: 56px;
      border-radius: 50%;
      background-color: var(--sendbird-light-background-50);
    }
  }
}

.sendbird-thumbnail-message-item-body:hover {
  cursor: pointer;
  .sendbird-thumbnail-message-item-body__image-cover {
    display: inline-flex;
  }
}
