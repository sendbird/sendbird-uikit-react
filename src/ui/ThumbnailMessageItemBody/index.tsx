import React, { ReactElement } from 'react';
import { FileMessage } from 'sendbird';
import './index.scss';

import Icon, { IconTypes, IconColors } from '../Icon';
import ImageRenderer from '../ImageRenderer';
import { getClassName, isGifMessage, isVideoMessage } from '../../utils';

interface Props {
  className?: string | Array<string>;
  message: FileMessage;
  isByMe?: boolean;
  mouseHover?: boolean;
  showFileViewer?: (bool: boolean) => void;
}

export default function ThumbnailMessageItemBody({
  className,
  message,
  isByMe = false,
  mouseHover = false,
  showFileViewer,
}: Props): ReactElement {
  const { thumbnails = [] } = message;
  const thumbnailUrl: string = thumbnails.length > 0 ? thumbnails[0]?.url : '';

  return (
    <div
      className={getClassName([
        className,
        'sendbird-thumbnail-message-item-body',
        isByMe ? 'outgoing' : 'incoming',
        mouseHover ? 'mouse-hover' : '',
        message?.reactions?.length > 0 ? 'reactions' : '',
      ])}
      onClick={() => showFileViewer(true)}
    >
      <ImageRenderer
        className="sendbird-thumbnail-message-item-body__thumbnail"
        url={thumbnailUrl || message?.url}
        alt={message?.type}
        width="360px"
        height="270px"
        placeHolder={(style) => (
          <div
            className="sendbird-thumbnail-message-item-body__placeholder"
            style={style}
          >
            <div className="sendbird-thumbnail-message-item-body__placeholder__icon">
              <Icon
                type={isVideoMessage(message) ? IconTypes.PLAY : IconTypes.PHOTO}
                fillColor={IconColors.ON_BACKGROUND_2}
                width="34px"
                height="34px"
              />
            </div>
          </div>
        )}
      />
      {
        (isVideoMessage(message) && !thumbnailUrl) && (
          <video className="sendbird-thumbnail-message-item-body__video">
            <source src={message?.url} type={message?.type} />
          </video>
        )
      }
      <div className="sendbird-thumbnail-message-item-body__image-cover" />
      {
        (isVideoMessage(message) || isGifMessage(message)) && (
          <div className="sendbird-thumbnail-message-item-body__icon-wrapper">
            <div className="sendbird-thumbnail-message-item-body__icon-wrapper__icon">
              <Icon
                type={isVideoMessage(message) ? IconTypes.PLAY : IconTypes.GIF}
                fillColor={IconColors.ON_BACKGROUND_2}
                width="34px"
                height="34px"
              />
            </div>
          </div>
        )
      }
    </div>
  );
}
