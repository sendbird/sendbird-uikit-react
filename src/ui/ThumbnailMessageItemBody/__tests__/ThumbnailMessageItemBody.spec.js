import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import ThumbnailMessageItemBody from "../index";

const createMockMessage = (process) => {
  const mockMessage = {
    url: 'https://static.sendbird.com/sample/user_sdk/user_sdk_25.png',
    name: 'A flag on the moon',
    type: 'image/png',
    messageType: 'file',
    reactions: [],
    thumbnails: [],
    isFileMessage: () => true,
    isUserMessage: () => false,
  };
  return process ? process(mockMessage) : mockMessage;
};

describe('ThumbnailMessageItemBody', () => {
  it('should have class names by own basic status', () => {
    const className = 'test-class-name';
    const component = mount(
      <ThumbnailMessageItemBody
        className={className}
        message={createMockMessage()}
      />
    );
    expect(
      component.find('.sendbird-thumbnail-message-item-body').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find(`.${className}.sendbird-thumbnail-message-item-body`).hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-thumbnail-message-item-body.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-thumbnail-message-item-body.incoming').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-thumbnail-message-item-body.mouse-hover').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-thumbnail-message-item-body.reactions').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-thumbnail-message-item-body__thumbnail').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-thumbnail-message-item-body__placeholder').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-thumbnail-message-item-body__video').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-thumbnail-message-item-body__image-cover').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-thumbnail-message-item-body__icon-wrapper').hostNodes().exists()
    ).toBe(false);

    expect(
      component.find('.sendbird-icon-photo').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('sendbird-icon-play').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('sendbird-icon-gif').hostNodes().exists()
    ).toBe(false);
  });

  it('should have class name by isByMe prop', () => {
    const outgoingMessage = mount(
      <ThumbnailMessageItemBody
        message={createMockMessage()}
        isByMe
      />
    );
    expect(
      outgoingMessage.find('.sendbird-thumbnail-message-item-body.outgoing').hostNodes().exists()
    ).toBe(true);
    expect(
      outgoingMessage.find('.sendbird-thumbnail-message-item-body.incoming').hostNodes().exists()
    ).toBe(false);
    const incomingMessage = mount(
      <ThumbnailMessageItemBody
        message={createMockMessage()}
        isByMe={false}
      />
    );
    expect(
      incomingMessage.find('.sendbird-thumbnail-message-item-body.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      incomingMessage.find('.sendbird-thumbnail-message-item-body.incoming').hostNodes().exists()
    ).toBe(true);
  });

  it('should have class name by mouseHover prop', () => {
    const component = mount(
      <ThumbnailMessageItemBody
        message={createMockMessage()}
        mouseHover
      />
    );
    expect(
      component.find('.sendbird-thumbnail-message-item-body.mouse-hover').hostNodes().exists()
    ).toBe(true);
  });

  it('should render video components when file type is video', () => {
    const component = mount(
      <ThumbnailMessageItemBody
        message={createMockMessage((mock) => ({
          ...mock,
          type: 'video/mp4',
        }))}
      />
    );
    expect(
      component.find('.sendbird-thumbnail-message-item-body__thumbnail').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-thumbnail-message-item-body__video').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-thumbnail-message-item-body__icon-wrapper').hostNodes().exists()
    ).toBe(true);

    expect(
      component.find('.sendbird-icon-photo').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-icon-play').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-icon-gif').hostNodes().exists()
    ).toBe(false);
  });

  it('should render gif components when file type is image/gif', () => {
    const component = mount(
      <ThumbnailMessageItemBody
        message={createMockMessage((mock) => ({
          ...mock,
          type: 'image/gif',
        }))}
      />
    );
    expect(
      component.find('.sendbird-thumbnail-message-item-body__thumbnail').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-thumbnail-message-item-body__video').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-thumbnail-message-item-body__icon-wrapper').hostNodes().exists()
    ).toBe(true);

    expect(
      component.find('.sendbird-icon-photo').hostNodes().exists()
    ).toBe(true); // as a default component
    expect(
      component.find('.sendbird-icon-play').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-icon-gif').hostNodes().exists()
    ).toBe(true);
  });

  it('should do a snapshot test of the ThumbnailMessageItemBody DOM', function() {
    const component = renderer.create(
      <ThumbnailMessageItemBody
        className="classname-for-snapshot"
        message={createMockMessage((mock) => ({
          ...mock,
          reactions: [{}, {}],
          thumbnails: ['https://static.sendbird.com/sample/user_sdk/user_sdk_25.png'],
        }))}
        isByMe
        mouseHover
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
