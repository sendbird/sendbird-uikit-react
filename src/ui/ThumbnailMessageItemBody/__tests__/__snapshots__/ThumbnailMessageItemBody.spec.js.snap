// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThumbnailMessageItemBody should do a snapshot test of the ThumbnailMessageItemBody DOM 1`] = `
<div
  className="classname-for-snapshot sendbird-thumbnail-message-item-body outgoing mouse-hover reactions"
  onClick={[Function]}
>
  <div
    className="sendbird-thumbnail-message-item-body__thumbnail sendbird-image-renderer"
    style={
      Object {
        "height": "270px",
        "maxWidth": "400px",
        "minWidth": "360px",
        "width": "100%",
      }
    }
  >
    <div
      className="sendbird-thumbnail-message-item-body__placeholder"
      style={
        Object {
          "style": Object {
            "alignItems": "center",
            "display": "flex",
            "height": "270px",
            "justifyContent": "center",
            "maxWidth": "400px",
            "minWidth": "360px",
            "position": "absolute",
            "width": "100%",
          },
        }
      }
    >
      <div
        className="sendbird-thumbnail-message-item-body__placeholder__icon"
      >
        <div
          className=" sendbird-icon sendbird-icon-photo sendbird-icon-color--on-background-2"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          style={
            Object {
              "height": "34px",
              "minHeight": "34px",
              "minWidth": "34px",
              "width": "34px",
            }
          }
          tabIndex="0"
        >
          <test-file-stub />
        </div>
      </div>
    </div>
    <div
      className="sendbird-image-renderer__image"
      style={
        Object {
          "backgroundImage": "url(https://static.sendbird.com/sample/user_sdk/user_sdk_25.png)",
          "backgroundPosition": "center",
          "backgroundRepeat": "no-repeat",
          "backgroundSize": "cover",
          "borderRadius": null,
          "height": "270px",
          "maxWidth": "400px",
          "minWidth": "360px",
          "position": "absolute",
          "width": "100%",
        }
      }
    />
    <img
      alt="image/png"
      className="sendbird-image-renderer__hidden-image-loader"
      onError={[Function]}
      onLoad={[Function]}
      src="https://static.sendbird.com/sample/user_sdk/user_sdk_25.png"
    />
  </div>
  <div
    className="sendbird-thumbnail-message-item-body__image-cover"
  />
</div>
`;
